vms/platformvm/state/subnet_only_validator.go

```diff
@@ -94,3 +166,129 @@ func putSubnetOnlyValidator(db database.KeyValueWriter, vdr *SubnetOnlyValidator
+	modifiedTotalWeight map[ids.ID]uint64 // subnetID -> totalWeight
+	modified            map[ids.ID]SubnetOnlyValidator
+	modifiedHasNodeIDs  map[subnetIDNodeID]bool
+	active              *btree.BTreeG[SubnetOnlyValidator]
+}
+
+func newSubnetOnlyValidatorsDiff() *subnetOnlyValidatorsDiff {
+	return &subnetOnlyValidatorsDiff{
+		modifiedTotalWeight: make(map[ids.ID]uint64),
+		modified:            make(map[ids.ID]SubnetOnlyValidator),
+		modifiedHasNodeIDs:  make(map[subnetIDNodeID]bool),
+		active:              btree.NewG(defaultTreeDegree, SubnetOnlyValidator.Less),
+	}
+}
+
+func (d *subnetOnlyValidatorsDiff) getActiveSubnetOnlyValidatorsIterator(parentIterator iterator.Iterator[SubnetOnlyValidator]) iterator.Iterator[SubnetOnlyValidator] {
+	return iterator.Merge(
+		SubnetOnlyValidator.Less,
+		iterator.Filter(parentIterator, func(sov SubnetOnlyValidator) bool {
+			_, ok := d.modified[sov.ValidationID]
+			return ok
+		}),
+		iterator.FromTree(d.active),
+	)
+}
+
+func (d *subnetOnlyValidatorsDiff) hasSubnetOnlyValidator(subnetID ids.ID, nodeID ids.NodeID) (bool, bool) {
+	subnetIDNodeID := subnetIDNodeID{
+		subnetID: subnetID,
+		nodeID:   nodeID,
+	}
+	has, modified := d.modifiedHasNodeIDs[subnetIDNodeID]
+	return has, modified
+}
+
+func (d *subnetOnlyValidatorsDiff) putSubnetOnlyValidator(state SubnetOnlyValidators, sov SubnetOnlyValidator) error {
+	var (
+		prevWeight uint64
+		prevActive bool
+		newActive  = sov.Weight != 0 && sov.EndAccumulatedFee != 0
+	)
+	switch priorSOV, err := state.GetSubnetOnlyValidator(sov.ValidationID); err {
+	case nil:
+		if !priorSOV.validateConstants(sov) {
+			return ErrMutatedSubnetOnlyValidator
+		}
+
+		prevWeight = priorSOV.Weight
+		prevActive = priorSOV.EndAccumulatedFee != 0
+	case database.ErrNotFound:
+		has, err := state.HasSubnetOnlyValidator(sov.SubnetID, sov.NodeID)
+		if err != nil {
+			return err
+		}
+		if has {
+			return ErrDuplicateSubnetOnlyValidator
+		}
+	default:
+		return err
+	}
+
+	switch {
+	case prevWeight < sov.Weight:
+		weight, err := state.WeightOfSubnetOnlyValidators(sov.SubnetID)
+		if err != nil {
+			return err
+		}
+
+		weight, err = safemath.Add(weight, sov.Weight-prevWeight)
+		if err != nil {
+			return err
+		}
+
+		d.modifiedTotalWeight[sov.SubnetID] = weight
+	case prevWeight > sov.Weight:
+		weight, err := state.WeightOfSubnetOnlyValidators(sov.SubnetID)
+		if err != nil {
+			return err
+		}
+
+		weight, err = safemath.Sub(weight, prevWeight-sov.Weight)
+		if err != nil {
+			return err
+		}
+
+		d.modifiedTotalWeight[sov.SubnetID] = weight
+	}
+
+	switch {
+	case prevActive && !newActive:
+		d.numAddedActive--
+	case !prevActive && newActive:
+		d.numAddedActive++
+	}
+
+	if prevSOV, ok := d.modified[sov.ValidationID]; ok {
+		prevSubnetIDNodeID := subnetIDNodeID{
+			subnetID: prevSOV.SubnetID,
+			nodeID:   prevSOV.NodeID,
+		}
+		d.modifiedHasNodeIDs[prevSubnetIDNodeID] = false
+		d.active.Delete(prevSOV)
+	}
+	d.modified[sov.ValidationID] = sov
```

Comment:
It might be true that:
```golang
prevSubnetIDNodeID := subnetIDNodeID{
	subnetID: prevSOV.SubnetID,
	nodeID:   prevSOV.NodeID,
}
d.modifiedHasNodeIDs[prevSubnetIDNodeID] = false
```
is dead code... The `d.active.Delete(prevSOV)` is needed because the tree is going to delete the entry based on the `EndAccumulatedFee` which may be different between `prevSoV` and `sov`.

Reply:
keeping this comment as a reminder for the deadcode. otherwise we can resolve this.